src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,115 @@
+import _ from 'lodash'
+import { timestamp } from './timestamp'
+import { pathToContext } from './pathToContext'
+import { timeDifference } from '../util.js'
+
+const editTimeMax = 30 // time diff limit in second for replacing descendants by ancestor
+
+const findTreeDeepestSubcontext = (tree, path) => {
+  if (path.length === 0) return { node: {}, path: [] }
+  const availableNode = _.get(tree, path)
+  if (availableNode) return { node: availableNode, path }
+  const pathIndex = path.findIndex((value, index) => _.get(tree, (path.slice(0, path.length - index))))
+  return pathIndex > -1 ? { node: _.get(tree, path.slice(0, path.length - pathIndex)), path: path.slice(0, path.length - pathIndex) } : {}
+}
+
+export const findTreeDescendants = (tree, startingPath, leafNodes = []) => {
+  const node = _.get(tree, startingPath)
+  if (!node) return []
+  if (node.leaf) leafNodes.push({ ...node }) // eslint-disable-line fp/no-mutating-methods
+  else Object.keys(node).forEach(child => findTreeDescendants(tree, [...startingPath, child], leafNodes))
```

Comment:
Let's use `reduce` here instead of mutation.

Reply:
Yes it makes sense.